pkg/traceql/engine_metrics.go

```diff
@@ -415,25 +523,42 @@ func (e *Engine) CompileMetricsQueryRange(req *tempopb.QueryRangeRequest, dedupe
 	me.start = req.Start
 	me.end = req.End
 
-	optimize(storageReq)
-
-	// Setup second pass callback.  Only as needed.
-	if len(storageReq.SecondPassConditions) > 0 {
-		storageReq.SecondPass = func(s *Spanset) ([]*Spanset, error) {
-			// The traceql engine isn't thread-safe.
-			// But parallelization is required for good metrics performance.
-			// So we do external locking here.
-			me.mtx.Lock()
-			defer me.mtx.Unlock()
-			return eval([]*Spanset{s})
-		}
+	// Setup second pass callback.  It might be optimized away
+	storageReq.SecondPass = func(s *Spanset) ([]*Spanset, error) {
+		// The traceql engine isn't thread-safe.
+		// But parallelization is required for good metrics performance.
+		// So we do external locking here.
+		me.mtx.Lock()
+		defer me.mtx.Unlock()
+		return eval([]*Spanset{s})
 	}
 
+	optimize(storageReq)
+
 	return me, nil
 }
 
 // optimize numerous things within the request that is specific to metrics.
 func optimize(req *FetchSpansRequest) {
+	if !req.AllConditions {
+		return
+	}
+
+	// There is an issue where multiple conditions &&'ed on the same
```

Comment:
This fixes a pre-existing unrelated bug.  Queries asserting multiple conditions on the same attribute like `span.http.status_code>=500 && span.http.status_code<600`, which should only fetch spans between 500 and 599, actually fetches them all since the conditions are checked independently, and the second pass engine logic filters them to the correct output. It means in this case we can't optimize away the second pass like for other metrics queries.  Ideally this is fixed in the Fetch layer and we bring the optimization back, but it wasn't trivial, so better to incur overhead instead of wrong results.